File: backend/core/analytics.py
# core/analytics.py
# Pluggable server-side product analytics (PostHog) — mirrors the
# AI_PROVIDER=mock / EMAIL_PROVIDER=mock pattern in core/config.py and
# core/email.py. POSTHOG_API_KEY unset means every call below is a no-op;
# the app behaves identically to before this existed. Server-side capture is
# the authoritative source for events tied to backend business logic (AI
# request completion + duration, signup, login) — it can't be blocked by an
# ad-blocker the way client-side capture can.

import logging
from functools import lru_cache

from core.config import settings

logger = logging.getLogger(__name__)


@lru_cache()
def _client():
    if not settings.posthog_api_key:
        return None
    from posthog import Posthog
    return Posthog(project_api_key=settings.posthog_api_key, host=settings.posthog_host)


def capture(distinct_id: str | int, event: str, properties: dict | None = None) -> None:
    client = _client()
    if client is None:
        logger.debug("ANALYTICS [no-op — set POSTHOG_API_KEY to enable] %s: %s %r", distinct_id, event, properties)
        return
    client.capture(event, distinct_id=str(distinct_id), properties=properties)


def identify(distinct_id: str | int, properties: dict | None = None) -> None:
    client = _client()
    if client is None:
        logger.debug("ANALYTICS [no-op — set POSTHOG_API_KEY to enable] identify %s: %r", distinct_id, properties)
        return
    client.set(distinct_id=str(distinct_id), properties=properties)

File: frontend/src/services/analytics.js
// services/analytics.js — Pluggable product analytics (PostHog). Mirrors the
// no-op-until-configured pattern already used for AI_PROVIDER/EMAIL_PROVIDER
// on the backend: VITE_POSTHOG_KEY unset means every call below is a no-op
// and the app behaves identically to before this existed.
import posthog from 'posthog-js'

const KEY  = import.meta.env.VITE_POSTHOG_KEY
const HOST = import.meta.env.VITE_POSTHOG_HOST || 'https://us.i.posthog.com'

let initialized = false

export function init() {
  if (initialized) return
  if (!KEY) {
    if (import.meta.env.DEV) {
      console.error('VITE_POSTHOG_KEY variable required by PostHog is missing or un-configured, this causes events to be silently missed. This error stops appearing once VITE_POSTHOG_KEY is configured')
    }
    return
  }
  posthog.init(KEY, {
    api_host: HOST,
    // Route changes are captured explicitly (see RouteTracker in main.jsx) —
    // this SPA never does a full page load between routes, so PostHog's own
    // autocapture pageview (which fires once, on script load) would miss
    // every navigation after the first.
    capture_pageview: false,
    // Session replay is opt-in per authenticated session (started/stopped
    // around login/logout in AuthContext), not left running on the public
    // marketing pages — see contexts/AuthContext.jsx.
    disable_session_recording: true,
    session_recording: {
      maskAllInputs: true,
    },
  })
  initialized = true
}

export function capture(event, properties = {}) {
  if (!initialized) return
  posthog.capture(event, properties)
}

export function identify(user) {
  if (!initialized || !user) return
  posthog.identify(String(user.id), { email: user.email, name: user.name })
}

export function reset() {
  if (!initialized) return
  posthog.stopSessionRecording()
  posthog.reset()
}

export function startSessionRecording() {
  if (!initialized) return
  posthog.startSessionRecording()
}

export function captureException(error, properties = {}) {
  if (!initialized) return
  posthog.captureException(error, { ...properties })
}
